src/LuaTransformer.ts

```diff
@@ -136,25 +160,51 @@ export class LuaTransformer {
             this.popScope();
 
             if (this.isModule) {
-                // If export equals was not used. Create the exports table.
-                // local exports = {}
-                if (!this.visitedExportEquals) {
-                    statements.unshift(
-                        tstl.createVariableDeclarationStatement(
-                            this.createExportsIdentifier(),
-                            tstl.createTableExpression()
-                        )
-                    );
+                if (!this.isWithinBundle) {
+                    // If export equals was not used. Create the exports table.
+                    // local exports = {}
+                    if (!this.visitedExportEquals) {
+                        statements.unshift(
+                            tstl.createVariableDeclarationStatement(
+                                this.createExportsIdentifier(),
+                                tstl.createTableExpression()
+                            )
+                        );
+                    }
                 }
 
                 // return exports
                 statements.push(tstl.createReturnStatement([this.createExportsIdentifier()]));
             }
         }
 
+        if (this.isWithinBundle) {
+            const moduleTableIdentifier = tstl.createIdentifier("____modules");
+            const exportPath = tsHelper.getExportPath(sourceFile.fileName, this.options);
+            const moduleParameters = this.visitedExportEquals ? undefined : [this.createExportsIdentifier()];
+            const moduleDeclaration = tstl.createAssignmentStatement(
+                tstl.createTableIndexExpression(moduleTableIdentifier, tstl.createStringLiteral(exportPath)),
+                tstl.createFunctionExpression(tstl.createBlock(statements, sourceFile), moduleParameters)
+            );
+            return tstl.createBlock([moduleDeclaration], sourceFile);
+        }
+
         return tstl.createBlock(statements, sourceFile);
     }
 
+    private transformSourceFileWithState(node: ts.SourceFile): tstl.Block {
```

Comment:
I think it would be a bit clearer to move it above `transformSourceFile`, since that's an internal method